Adding truncated octahedron model (python and c)

sara-mokhtari opened the pull request:
## Description

Adding a new model for truncated octahedrons (python + C).

Fixes #684 


## How Has This Been Tested?

Unit tests worked. 
Other tests were perfomed to validate the model: for truncated octahedrons with different sizes and truncatures, agreement was found with numerical calculation using Debye equation (Debye calculator: https://github.com/FrederikLizakJohansen/DebyeCalculator ).  It shows good agreement between the two numerical techniques (Debye based on atomic positions and SasView model based on analytical expressions).

More tests should be made on small q. Indeed, we encouter issues when it comes to small q.  More precise mathematical should be used (hospital rule, etc).


## Review Checklist:

[if using the editor, use `[x]` in place of `[ ]` to check a box]

**Documentation** (check at least one)
- [ ] There is **nothing** that needs documenting
- [x] Documentation changes are **in this PR**
- [ ] There is an **issue** open for the documentation (link?)

**Installers**
- [ ] There is a chance this will affect the **installers**, if so
  - [ ] **Windows** installer (GH artifact) has been tested (installed and worked) 
  - [ ] **MacOSX** installer (GH artifact) has been tested (installed and worked)
  - [ ] **Wheels** installer (GH artifact) has been tested (installed and worked)

**Licensing** (untick if necessary)
- [x] The introduced changes comply with SasView license (BSD 3-Clause)

marimperorclerc commented:
This is a great start for us !!! :) About issues at small q, it is probably because no special treatment for singularities is included in the code. Singularities (i.e. division by zero) are happening for specific directions depending on the shape of the polyhedron (along vertices, edges and normal to the facets). So far, in what we have implemented, the code is running well because all quadrature points are not along any singularity direction. But a better solution would be a nice improvement, by taking into account singularity directions explicitely in the analytical expression of the form factor.

pkienzle commented on sasmodels/models/octahedron_truncated.py:
```diff
@@ -0,0 +1,232 @@
+    q_z = q\,\cos\theta
+
+θ is the angle between the scattering vector and the z axis of the octahedron (length_c).
+ϕ is the angle between the scattering vector (lying in the xy plane) and the x axis (length_a).
```
Other models are using (a, b, c) as the axes for the shape rather than (x, y, z).

wpotrzebowski commented on sasmodels/models/octahedron_truncated.c:
```diff
@@ -0,0 +1,226 @@
+#include <math.h>
+#include <stdio.h> 
+static double
+form_volume(double length_a, double b2a_ratio, double c2a_ratio, double t)
```
Thanks a lot for the contribution! 
I am not really expert on these models but it seems to me there are t values that entire volume can go below zero, which may cause problems along the way.

wpotrzebowski replied:
The simple solution to it, may be to contsrain t  parameter in the parameter table.

wpotrzebowski replied:
Thank you for explaining. I've been simply looking at this term (1.-3*(1.-t)*(1.-t)*(1.-t)) and where it zeros.

wpotrzebowski commented on sasmodels/models/octahedron_truncated.c:
```diff
@@ -0,0 +1,226 @@
+    	    const double qy = Qy * length_b;
+    	    const double qz = Qz * length_c;
+
+            const double AA = 1./((qy*qy-qz*qz)*(qy*qy-qx*qx))*((qy-qx)*sin(qy*(1.-t)-qx*t)+(qy+qx)*sin(qy*(1.-t)+qx*t))+
```
(qy*qy - qz*qz) probably needs to be taken care of. Otherwise we can end up in NaN/Inf scenario

wpotrzebowski commented on sasmodels/models/octahedron_truncated.c:
```diff
@@ -0,0 +1,226 @@
+
+    // Multiply by contrast^2 and volume^2
+    // volume of octahedron
+    const double volume = (4./3.)*length_a * length_b * length_c*(1.-3*(1.-t)*(1.-t)*(1.-t));
```
Isn't it the same as form_volume?

wpotrzebowski replied:
If yes, better reuse the function

sara-mokhtari commented on sasmodels/models/octahedron_truncated.py:
```diff
@@ -0,0 +1,232 @@
+.. math::
+    q_x = q\,\sin\theta\,\cos\phi, \qquad
+    q_y = q\,\sin\theta\,\sin\phi, \qquad
+    q_z = q\,\cos\theta
```
We changed into AA, BB, CC.

sara-mokhtari commented:
We updated the codes (both octahedron_truncated.c and octahedron_trucated.py)  with more specific documentation and fixed some mistakes. We also added a new image that describes better the truncation parameter of the model. 
Feel free to review it again. We explained some of our choices in the previous conversations. If you see remaining issues, let us know. Thank you !

pkienzle commented on sasmodels/models/octahedron_truncated.py:
```diff
@@ -0,0 +1,253 @@
+    [{"background": 0, "scale": 1, "length_a": 100, "t": 1, "sld": 1., "sld_solvent": 0.},
+     0.01, 120.57218749910827],
+    [{"background": 0, "scale": 1, "length_a": 100, "t": 1, "sld": 1., "sld_solvent": 0.},
+     [0.01, 0.1], [120.57218749910827, 0.3741758252985113]],
```
Please note the source of the target values.

Make sure you are using an appropriate number of sigfigs on the target value. For example, if your target values come from a numerical integral only report the digits supported by the integration stopping conditions.

The sasmodels test code is only looking at the first five sigfigs of the target during the comparison. You may include more but for now we won't check them.

You could add a truncated octahedron to shape2sas if you wanted a further source of validation values.

marimperorclerc replied:
To @pkienzle:

About validation, for information, here we are sharing the numerical validation we have performed by comparing with Debye-calculator for different truncated octahedrons:
<img width="2978" height="977" alt="trOh_trunc_sasview_debye" src="https://github.com/user-attachments/assets/05e50f29-61d1-443e-828b-4ef716ed90ec" />

This is not yet published but it is a strong validation of the analytical model as the calculation method based on atomic coordinates is completely different.

And thank you for mentioning shape2sas !

About the source of the target value, it is simply obtained by running the model in SASView. We will keep only 5 digits in the test block at the end of the python file.

marimperorclerc replied:
5 digits are OK with us.
